### PR TITLE
Varnish HTTP Purge v 4.1

### PR DESCRIPTION
**Pull request by Ipstenu**
**Major Release**

The biggest part of this is that we're supporting the REST API more fully. However we're also dropping pre-4.7 support in doing so (the older versions will work on older WP). Two pull requests for filters by trusted WP community peoples are included, as well as support for Private pages (not an issue on DreamPress but possibly for others) and a lot of code formatting cleanup.

* JSON / REST API Support
* Fix for Varnish Status Page on MAMP (props @jeremyclarke)
* Filter for purge headers (props @ocean90)
* Disallow people from editing the Varnish IP on Multisite
* Drop support for pre 4.7 because of JSON / REST API
* Support flushing cache for private pages

### CODE REVIEW

File: varnish-http-purge.php [hunk 1]
```diff
@@ -347,94 +380,131 @@ public function purgeNoID( $postId ) {
 	 */
 	public function purgePost( $postId ) {
 		
+		global $wp_version;
```
Comment (ocean90) [inline varnish-http-purge.php]: Seems like the global isn't used in this method.

File: varnish-http-purge.php [hunk 2]
```diff
@@ -347,94 +380,131 @@ public function purgeNoID( $postId ) {
...
+
+			// Post URL
+			array_push( $listofurls, get_permalink( $postId ) );
+
+			// JSON API Permalink for the post based on type
+			// We only want to do this if the rest_base exists
+			// But we apparently have to force it for posts and pages (seriously?)
+			$post_type_object = get_post_type_object( $postId );	
+			if ( isset( $post_type_object->rest_base ) ) {
+				$rest_permalink = get_rest_url().$rest_api_route.'/'.$post_type_object->rest_base.'/'.$postId.'/';
```
Comment (ocean90) [inline varnish-http-purge.php]: `rest_url( $rest_api_route . '/' . $post_type_object->rest_base . '/' . $postId . '/' )` looks much better IMO. Does the REST API really use trailing slashes here?

Comment (Ipstenu) [inline varnish-http-purge.php]: The REST API does use trailing slashes pretty much everywhere. Both `http://plugins.wordpress.dev/wp-json/wp/v2/posts/1/` and `http://plugins.wordpress.dev/wp-json/wp/v2/posts/1` work, mind you, but the same isn't true of all add-ons. This proved to be more robust when testing with weird things people do in the real world.

File: varnish-http-purge.php (hunk 1 above)
Comment (Ipstenu) [inline varnish-http-purge.php]: Hah, that's a remnant from when I was spec'ing out how I wanted to handle 'what if WP changes the JSON versions!!!'

File: varnish-http-purge.php (hunk 2 above)
Comment (ocean90) [inline varnish-http-purge.php]: Right it works, but if you take a look at https://wordpress.org/news/wp-json/wp/v2/posts and the `_links` property you'll notice that no trailing slash is appended by default.

Comment (ocean90) [inline varnish-http-purge.php]: btw: I wasn't actually complaining the (missing) whitespace usage. :) More about rest_url() vs. get_rest_url().

Comment (Ipstenu) [inline varnish-http-purge.php]: One day I'll understand why GitHub doesn't email me for your replies on this... get_rest_url uses blog_id which means Multisite shenanigans are less so. Which is why I used it :)

Comment (ocean90) [inline varnish-http-purge.php]: @Ipstenu But you don't pass any `blog_id`s to `get_rest_url()` thus it's the same as `rest_url()`. 🤔

**Pull request merged**